dev/TEAL_opcodes.md

```diff
@@ -0,0 +1,440 @@
+
+- Opcode: 0x01 
+- Pops: *... stack*, []byte
+- Pushes: []byte
+- SHA256 hash of value, yields [32]byte
+- **Cost**: 7
+
+## keccak256
+
+- Opcode: 0x02 
+- Pops: *... stack*, []byte
+- Pushes: []byte
+- Keccac256 hash of value, yields [32]byte
+- **Cost**: 26
+
+## sha512_256
+
+- Opcode: 0x03 
+- Pops: *... stack*, []byte
+- Pushes: []byte
+- SHA512_256 hash of value, yields [32]byte
+- **Cost**: 9
+
+## ed25519verify
+
+- Opcode: 0x04 
+- Pops: *... stack*, {[]byte A}, {[]byte B}, {[]byte C}
+- Pushes: uint64
+- for (data, signature, pubkey) verify the signature of the data against the pubkey => {0 or 1}
+- **Cost**: 1900
+
+## rand
+
+- Opcode: 0x05 
+- Pops: _None_
+- Pushes: uint64
+- push random uint64 to stack
+- **Cost**: 3
+
+Random number generator based on the ChaCha20 algorithm. Seeded with the previous block's `Seed` value and the current transaction ID.
+
+## +
+
+- Opcode: 0x08 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A plus B. Panic on overflow.
+
+## -
+
+- Opcode: 0x09 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A minus B. Panic if B > A.
+
+## /
+
+- Opcode: 0x0a 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A divided by B. Panic if B == 0.
+
+## *
+
+- Opcode: 0x0b 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A times B. Panic on overflow.
+
+It is worth noting that there are 10,000,000,000,000,000 micro-Algos in the total supply, or a bit less than 2^54. When doing rational math, e.g. (A * (N/D)) as ((A * N) / D) one should limit the numerator to less than 2^10 to be completely sure there won't be overflow.
+
+## <
+
+- Opcode: 0x0c 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A less than B => {0 or 1}
+
+## >
+
+- Opcode: 0x0d 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A greater than B => {0 or 1}
+
+## <=
+
+- Opcode: 0x0e 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A less than or equal to B => {0 or 1}
+
+## >=
+
+- Opcode: 0x0f 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A greater than or equal to B => {0 or 1}
+
+## &&
+
+- Opcode: 0x10 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A is not zero and B is not zero => {0 or 1}
+
+## ||
+
+- Opcode: 0x11 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A is not zero or B is not zero => {0 or 1}
+
+## ==
+
+- Opcode: 0x12 
+- Pops: *... stack*, {any A}, {any B}
+- Pushes: uint64
+- A is equal to B => {0 or 1}
+
+## !=
+
+- Opcode: 0x13 
+- Pops: *... stack*, {any A}, {any B}
+- Pushes: uint64
+- A is not equal to B => {0 or 1}
+
+## !
+
+- Opcode: 0x14 
+- Pops: *... stack*, uint64
+- Pushes: uint64
+- X == 0 yields 1; else 0
+
+## len
+
+- Opcode: 0x15 
+- Pops: *... stack*, []byte
+- Pushes: uint64
+- yields length of byte value
+
+## btoi
+
+- Opcode: 0x17 
+- Pops: *... stack*, []byte
+- Pushes: uint64
+- converts bytes as big endian to uint64
```

Comment:
Does this panic if the the length of the byte slice is higher than 8?

Reply:
yes, noted